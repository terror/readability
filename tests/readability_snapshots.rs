use {
  readability::{Readability, ReadabilityOptions},
  std::{
    fs,
    path::{Path, PathBuf},
  },
};

fn load_fixture(path: &Path) -> anyhow::Result<String> {
  Ok(fs::read_to_string(path.join("source.html"))?)
}

fn run_fixture(name: &str, path: PathBuf) -> anyhow::Result<()> {
  let html = load_fixture(&path)?;

  let mut parser =
    Readability::new(&html, None, ReadabilityOptions::default())?;

  let article = parser.parse()?;

  insta::with_settings!({ snapshot_suffix => name }, {
    insta::assert_json_snapshot!("article", &article);
  });

  Ok(())
}

fn fixture_dirs() -> Vec<(String, PathBuf)> {
  let mut fixtures = Vec::new();

  let root = PathBuf::from("tests/fixtures");

  if root.exists() {
    for entry in fs::read_dir(&root).expect("fixtures directory") {
      let entry = entry.expect("fixture entry");

      if entry.file_type().map(|ty| ty.is_dir()).unwrap_or(false) {
        let path = entry.path();

        if path.join("source.html").exists() {
          let name = entry.file_name().to_string_lossy().to_string();
          fixtures.push((name, path));
        }
      }
    }
  }

  if let Ok(env_root) = std::env::var("READABILITY_TEST_ROOT") {
    let env_path = PathBuf::from(env_root);

    if env_path.exists() {
      for entry in walkdir::WalkDir::new(env_path)
        .max_depth(2)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|e| e.file_type().is_dir())
      {
        let dir = entry.path().to_path_buf();

        if dir.join("source.html").exists() {
          let name = dir
            .file_name()
            .unwrap_or_default()
            .to_string_lossy()
            .to_string();

          fixtures.push((name, dir));
        }
      }
    }
  }

  fixtures
}

#[test]
fn readability_snapshots() -> anyhow::Result<()> {
  let fixtures = fixture_dirs();

  if fixtures.is_empty() {
    return Ok(());
  }

  for (name, path) in fixtures {
    run_fixture(&name, path)?;
  }

  Ok(())
}
