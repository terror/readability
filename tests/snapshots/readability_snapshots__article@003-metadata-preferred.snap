---
source: tests/readability_snapshots.rs
expression: "&article"
---
{
  "byline": "Dublin Core property author",
  "content": "<div id=\"readability-content\"><div id=\"readability-page-1\" class=\"page\"><article>\n      <h1>Test document title</h1>\n      <p>\n        Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut\n        labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco\n        laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in\n        voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat\n        cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\n      </p>\n      <p>\n        Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut\n        labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco\n        laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in\n        voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat\n        cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\n      </p>\n    </article></div></div>",
  "dir": null,
  "excerpt": "Dublin Core property description",
  "lang": null,
  "length": 913,
  "published_time": null,
  "site_name": null,
  "text_content": "Test document title Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum. Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
  "title": "Dublin Core property title"
}
